draft.md

```diff
@@ -0,0 +1,81 @@
+# Draft Ipld Unixfs Spec
+
+## Basic Structure
+
+   - Some sort of header that indicates that this a directory and included a version number.  The header could also have fields to give additional information on the meaning of the extended attributes.
+
+  - CBOR Map
+    - Key: CBOR Byte or Text String: File Name
+    - Value: CBOR Array of:
+      - Type: CBOR Unsigned Int
+      - Link or Data: CBOR Type varies
+      - Optional file size: CBOR Unsigned Int
+      - Optional Standard Attributes: CBOR Map
+      - Optional Extended Attributes: CBOR Map
+
+The file size is only defined for regular files and is the size of the file contents.
+
+All maps should be ordered based on the binary values of the key,
+duplicates are not allowed.
+
+### Notes
+
+ * An array makes sense to be as this is more compact and the value of
+   the fields are unambiguous, it also allows for a separation of
+   standard and extended attributes
+
+ * The key type can either be a byte or text string as POSIX makes no
+   requirements that file names be utf-8 and it is important that any
+   file name can be faithfully represented, if the string is utf-8
+   then the type will be Text.
+
+## Types
+
+The type field should be limited to a set of well defined values so it
+makes sense that this is an integer rather than a text string.  The
+value is the ascii value of a letter. When converting to JSON the
+integer can be represented as a single character string.
+
+Possible values are as follows:
+
+  * 0, '', `file`: regular file
+  * `e`, `exe`: executable file
+  * `d`, `dir`: directory entry
+  * `s`, `special`: special file type (fifo, device, etc). The second field is a CBOR Map with at least one field to describe the type.
+  * `l`, `symlink`: symbolic link.  The second field is the contents of the link
+  * `o`, `other`: link to other ipld object, links followed for GC and related operations 
+  * `u`, `unknown`: link to unknown objects, links not followed
+
+### Notes
+
+  * Rather than have a special attribute for an executable bit it is more compact if we just make this a different type
+  * It is very useful to be able to determine if a link is a directory or an ordinary file so I made it as separate type, also there can be multiple ways to define a file size for a directory so it is best to just leave it out as it is of limited usefulness
```

Comment:
> there can be multiple ways to define a file size for a directory

Actually there are only 2 ways - either you only count the logical bytes ( what the Windows `properties` UI does ), or you take into account the allocation overhead of the filesystem - the blocks taken by both the directories and the files themselves rounded up ( what the unixish `du` does ).

Given that in the context of IPFS the DAG is completely decoupled from the storage ( it may be files, it may be badger, etc ), the only sensible way to define a file size for a directory is to count the logical bytes, which I've done in my prototypes.

I would be sad if I can't express these cumulative values as part of every link within an FS tree.

Reply:
Unix has it's own (basically useless) way of defining the size of a directory.

I am not totally against included the cumulative size of a directory if we can agree on how to define it.